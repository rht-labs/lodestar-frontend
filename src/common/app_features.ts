// All roles defined by the backend to be used in the frontend should be enumerated in this object.

export const APP_FEATURES = {
  admin: 'admin',
  writer: 'writer',
  reader: 'reader',
  notifications: 'notifications',
  engagementCardIcons: 'engagementCardIcons',
  resetUser: 'resetUser',
};

export type AppFeature = keyof typeof APP_FEATURES;
