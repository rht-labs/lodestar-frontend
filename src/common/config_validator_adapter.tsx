import { ValidatorFactory } from '../schemas/validators';
import { EngagementFormConfig } from '../schemas/engagement_config';

export const getValidatorsFromFormOptions = (
  formOptions: EngagementFormConfig = {}
) =>
  Object.keys(formOptions || {}).reduce(
    (previousSchemaGroups, currentGroupKey) => {
      return {
        ...previousSchemaGroups,
        ...Object.keys(formOptions[currentGroupKey] ?? {}).reduce(
          (previousGroupFields, currentFieldKey) => ({
            ...previousGroupFields,
            [currentFieldKey]: (
              formOptions?.[currentGroupKey]?.[currentFieldKey]?.validators ||
              []
            ).map(ValidatorFactory),
          }),
          {}
        ),
      };
    },
    {}
  );
