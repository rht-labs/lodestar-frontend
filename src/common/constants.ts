export const APPLICATION_CONTENT_CONTAINER_ID = 'main-application-content';
