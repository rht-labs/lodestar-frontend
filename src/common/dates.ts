import { format as formatDate, utcToZonedTime } from 'date-fns-tz';
export const formatUtcDate = (date?: Date): string | undefined => {
  if (!date) {
    return '';
  }
  const dateString = formatDate(utcToZonedTime(date, 'UTC'), 'MMM dd, yyyy', {
    timeZone: 'UTC',
  });
  return dateString;
};
