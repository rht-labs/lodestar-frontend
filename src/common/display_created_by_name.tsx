import React from 'react';

export function DisplayCreatedByName({
  userFromServer,
  lastUpdatedBy,
}: {
  userFromServer?: string;
  lastUpdatedBy?: string;
}) {
  return <>{!!userFromServer ? userFromServer : lastUpdatedBy}</>;
}
