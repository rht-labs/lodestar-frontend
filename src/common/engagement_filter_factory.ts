import {
  EngagementFilter,
  EngagementSortFields,
} from '../schemas/engagement_filter';
import { Engagement, getEngagementStatus } from '../schemas/engagement';

function hasSearchTermMatch(engagement: Engagement, searchTerm: string = '') {
  if (!searchTerm || !engagement) {
    return true;
  }
  const cleanedSearchTerm = searchTerm?.trim()?.toLowerCase();
  const cleanedCustomerName = engagement.customer_name.toLowerCase().trim();
  const cleanedProjectName = engagement.project_name.toLowerCase().trim();
  return (
    cleanedProjectName.includes(cleanedSearchTerm) ||
    cleanedCustomerName.includes(cleanedSearchTerm)
  );
}

/**
 * The Engagement Filter Factory accepts a filter definition as a parameter.
 * It parses the filter definition and returns a single filter function that
 * filters an array of Engagements based on the terms of the filter definition.
 */
export const engagementFilterFactory = (
  filter?: EngagementFilter
): ((engagement: Engagement) => boolean) => {
  if (!filter || Object.keys(filter).length === 0) {
    return () => true;
  }
  return (engagement: Engagement) => {
    const filterTestResults: boolean[] = [];
    if (filter.allowedStatuses) {
      filterTestResults.push(
        filter.allowedStatuses.includes(getEngagementStatus(engagement))
      );
    }
    if (filter.searchTerm) {
      filterTestResults.push(
        hasSearchTermMatch(engagement, filter?.searchTerm)
      );
    }
    /**
     * If filterTestResult has a length of 0, this means no tests were performed.
     * The default behavior of the filter should be to show engagements.
     * Therefore, we return true
     */
    if (filterTestResults.length === 0) {
      return true;
    }
    /**
     * If filterTestResults has results, we only return if every result is true.
     */
    return filterTestResults.every(r => !!r);
  };
};

const engagementSorters = (sortOrderMultiplier: number = 1) => {
  return {
    startDate: (a: Engagement, b: Engagement) =>
      (a.start_date > b.start_date ? 1 : -1) * sortOrderMultiplier,

    endDate: (a: Engagement, b: Engagement) =>
      (a.end_date > b.end_date ? 1 : -1) * sortOrderMultiplier,

    customerName: (a: Engagement, b: Engagement) =>
      (a.customer_name > b.customer_name ? 1 : -1) * sortOrderMultiplier,

    projectName: (a: Engagement, b: Engagement) =>
      (a.project_name > b.project_name ? 1 : -1) * sortOrderMultiplier,
  };
};

/**
 * The Engagement Sort Factory parses the `sort` field of the Engagement filter definition.
 * It returns a single function that can be passed to the `sort` method of an Engagement array.
 */
export const engagementSortFactory = ({ sort }: EngagementFilter = {}): ((
  a: Engagement,
  b: Engagement
) => number) => {
  /**
   * The sort order multiplier translates the isAscending boolean field into
   * an integer that can invert the sort order of the array.
   */
  const sortOrderMultiplier =
    sort?.isAscending === undefined || sort?.isAscending ? 1 : -1;
  if (sort?.sortField === EngagementSortFields.startDate) {
    return engagementSorters(sortOrderMultiplier).startDate;
  }
  if (sort?.sortField === EngagementSortFields.endDate) {
    return engagementSorters(sortOrderMultiplier).endDate;
  }
  if (sort?.sortField === EngagementSortFields.customerName) {
    return engagementSorters(sortOrderMultiplier).customerName;
  }
  if (sort?.sortField === EngagementSortFields.projectName) {
    return engagementSorters(sortOrderMultiplier).projectName;
  }
  return () => 0;
};
