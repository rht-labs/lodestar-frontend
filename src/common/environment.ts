export class Environment {
  public getValue(key: string) {
    return process.env[key];
  }
}
