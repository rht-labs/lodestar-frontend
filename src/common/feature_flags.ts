//true means the flag is on - The feature is hidden
export const FEATURE_FLAG_NOTIFICATION = true;
export const FEATURE_FLAG_ICONS_IN_ENGAGEMENT_CARD = true;