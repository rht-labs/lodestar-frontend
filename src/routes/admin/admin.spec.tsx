import React from 'react';
import { Admin } from '.';
import { TestStateWrapper } from '../../common/test_state_wrapper';

describe('Admin UI', () => {
  test('should match snapshot', () => {
    expect(
      <TestStateWrapper>
        <Admin></Admin>
      </TestStateWrapper>
    ).toMatchSnapshot();
  });
});
