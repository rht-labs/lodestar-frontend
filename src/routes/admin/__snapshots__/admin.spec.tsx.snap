// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin UI should match snapshot 1`] = `
<TestStateWrapper>
  <UNDEFINED />
</TestStateWrapper>
`;
